File: 프로그래머스/lv3/43238. 입국심사/입국심사.py

import sys, copy, heapq
import heapq, math
from queue import PriorityQueue
from itertools import permutations, combinations, product
from collections import deque
from itertools import product

#sys.setrecursionlimit(10 ** 5)
#a=list(product(i,repeat=len(powers)))
#b=list(product(*a)) #리스트 안에 있는 원소들끼리 조합
#sys.stdin = open("input.txt","r")

from itertools import combinations_with_replacement as cwr
from collections import Counter
input = sys.stdin.readline



def binary(left, right, times, n):
    cnt = 0
    answer = -1

    while left <= right:
        mid = int((left + right) / 2)
        cnt = 0
        for i in times:
            cnt += int(mid / i)
        if cnt < n:

            left = mid + 1
        else :
            right = mid - 1

    return left


def solution(n, times):
    times.sort()
    left = 0
    right = times[-1]*n
    answer = binary(left, right, times, n)

    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;
class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        long left=-1;
        long right=-1;
        for(int i:times){
            right=Math.max(right,i);
        }
        right=right*n+1;
        
        while(left+1<right){
            long mid=(left+right)/2;
            long temp=0;
            for(int i: times){
                temp+=mid/i;
                if(temp>=n)
                    break;
            }
            
            if(temp>=n){
                right=mid;
                answer=mid;
            }
            else{
                left=mid;
            }
            
        }
            
        return answer;
    }
}